### ISSUE TITLE
hole_size_compensation applies to rectangles and squares

### ISSUE THREAD

**Issue by macchinette** [supermerill/SuperSlicer#1959]
**Describe the bug**
Hole_size_compensation applies to rectangles and squares

**To Reproduce**
1. Open 3mf attachment
2. Cut files
3. Scroll down to layer 177
4. See error boat cabin

**>> Project File <<**

[Hole_size_compensation.zip](https://github.com/supermerill/SuperSlicer/files/7604895/Hole_size_compensation.zip)

**Expected behavior**
Apply the function to round holes only?

**Screenshots**
![image](https://user-images.githubusercontent.com/79080486/143491124-bdc26d9f-51c4-48e0-aa24-355100642181.png)

**Desktop (please complete the following information):**
 - OS: Windows 10 Pro
 - Version SuperSlicer_2.3.57.6

**Additional context**
I know that using the hole_size_threshold parameter you can limit the trigger area, but if there are large holes then it also applies to smaller squares and rectangles than the holes.

**Comment by supermerill** [comment]
I guess i can add the same checks than for the polyholes.

**Comment by supermerill** [comment]
also, it may be useful o add a "gradient" for when the hole stop and then restart (when the vertical hole is broken open by a horizontal hole)

**Comment by macchinette** [comment]
> Immagino di poter aggiungere gli stessi controlli che per i polihole.

Thanks for everything you do!

**Comment by telepresencebot2** [comment]
Isn't this working as intended? It never claims to only apply to round holes.

**Comment by macchinette** [comment]
The operation popup indicates the opposite of how it actually works.  Then there is the bug that applies to all closed perimeters empty in the center as if they were round holes.

Inviato da iPhone

> Il giorno 13 ago 2022, alle ore 22:51, telepresencebot2 ***@***.***> ha scritto:
> 
> Isn't this working as intended? It never claims to only apply to round holes.

**Comment by telepresencebot2** [comment]
Yes, there is a separate ticket open for the tool tip being incorrect about the direction of compensation, but nothing about the tool tip or the name of the feature suggests it should only apply to round holes. It says it applies to interior perimeters.

**Comment by macchinette** [comment]
okay it is not specified both for round holes, but since the function to enlarge and narrow the internal perimeters already exists and it is XY compensation, where it says Vertical Hole shrinking compensation is very useful for all round holes which are normally smaller and it doesn't happen to me for all the other closed loop perimeters that aren't round, so if it only applied to round holes it would be a wonderful thing.

Inviato da iPhone

**Comment by telepresencebot2** [comment]
XY compensation applies to external perimeters. (It claims in its tool tip to only apply to exterior shells, have not tested though)
Vertical hole compensation is for internal perimeters.

**Comment by macchinette** [comment]
it is not true XY compensation also applies to the internal perimeters, I'm sure of it because I tried it, for what a function only for round holes is fundamental, because all the round holes in my case are smaller, but the internal perimeters of an empty piece inside no.

Inviato da iPhone

> Il giorno 15 ago 2022, alle ore 13:22, telepresencebot2 ***@***.***> ha scritto:
> 
> XY compensation applies to external perimeters. (It claims in its tool tip to only apply to exterior shells, have not tested though)
> Vertical hole compensation is for internal perimeters.

**Comment by bitingsock** [comment]
For clarification, is there currently any functional difference between "vertical hole shrinking compensation" and "XY compensation > inner"?

**Comment by telepresencebot2** [comment]
"xy compensation > inner" says in it's tool tip that it applies to inner shells. "xy compensation > outer" also states that it should probably always match "xy compensation > inner".

Also, "vertical hole compensation" allows changing the threshold for what is considered a hole and what isn't, "xy compensation > outer" does not.
